### PR TITLE
core: fix stack overflow when displaying passphrase keyboard

### LINKED ISSUE #2779: UI2: stack overflow in passphrase keyboard
**Describe the bug**
RSOD when assking for passphrase on device. Affects firmware but **not emulator**.

**Firmware version and revision**
master

**To Reproduce**
`trezorctl set passphrase on && trezorctl btc get-public-node -n 'm/44h/0h/0h'`

### PR DESCRIPTION
**Pull request by mmilata**
Fixes #2779, related to #2708.

Size of `PassphraseKeyboard`:
* original: 4340 B
* after 24bcb74: 3668 B
* after 3487811: 2996 B
* after 6a4d470: 956 B

Not implemented:
* placement new (=construct object directly on micropython heap): seems tricky at the moment, would probably be feasible with global allocator
* further refactoring of `Button`: it's possible to reduce memory usage even further but the gains are not as big but with more effort

### CODE REVIEW

File: core/embed/rust/src/ui/display/loader.rs [hunk 1]
```diff
@@ -337,7 +338,7 @@ pub fn loader_rust(
 
             icon_buffer_used.buffer[icon_offset as usize..(icon_offset + icon_width / 2) as usize]
                 .copy_from_slice(
-                    &unwrap!(icon_data).toif.data[(y_i * (icon_width / 2)) as usize
+                    &unwrap!(icon_data).toif.zdata()[(y_i * (icon_width / 2)) as usize
```
Comment (mmilata) [inline core/embed/rust/src/ui/display/loader.rs:341]: Don't we need to decompress the data here first? Are we even using loader with icon anywhere? Bootloader?

File: core/embed/rust/src/ui/display/toif.rs [hunk 2]
```diff
@@ -72,41 +67,46 @@ pub fn icon(icon: &Icon, center: Point, fg_color: Color, bg_color: Color) {
...
-            name: named_toif.1,
-            size: info.0,
-            format: info.1,
-        }
+    pub fn new(data: &'static [u8]) -> Self {
+        Self { data }
+    }
+
+    pub fn format(&self) -> ToifFormat {
+        let info = unwrap!(toif_info(self.data));
```
Comment (matejcik) [inline core/embed/rust/src/ui/display/toif.rs:79]: unconvinced about those -- we're calling `toif_info` all over the place uselessly.

it seems slightly better to just implement `display_toif_info` in Rust here:

1. `new()` will assert the TOIF header
2. `format()` will match on the format character
3. `width()` and `height()` will just parse out the respective bytes
4. `size()` will combine `width()` + `height()`

all of those _look like_ they could be `const fn`s too, with the possible exception of `new()`?

**Review by matejcik — APPROVED** [review]
LGTM, like it a lot. 

ACK on the passphrase keyboard change and on removing `NamedToif`

OK with the `Toif` changes. I would like to see the changes I mentioned in the comments, but we can proceed like this if you think that would be better solved in a separate issue.

File: core/embed/rust/src/ui/display/loader.rs (hunk 1 above)
Comment (TychoVrahe) [reply core/embed/rust/src/ui/display/loader.rs:341]: yes there is a bug in here, the `loader_rust` is actually meant to receive already decompressed data, which allows to skip repeated decompression when updating progress. 

in bootloader the icons are used, so i made a fix there when i noticed this after rebase, but maybe you can cherry pick it here: https://github.com/trezor/trezor-firmware/commit/e8bfaaf0ce8931678427c9ef304bce3f83feda70

Comment (mmilata) [reply core/embed/rust/src/ui/display/loader.rs:341]: Let's fix it in that PR if it's only used in bootloader.

File: core/embed/rust/src/ui/display/toif.rs (hunk 2 above)
Comment (mmilata) [reply core/embed/rust/src/ui/display/toif.rs]: Good idea, implemented in [9ecb7b0](https://github.com/trezor/trezor-firmware/pull/2792/commits/9ecb7b0e4bcd9437f6b9b763ff4a3238cad03e3d)

File: core/embed/rust/src/ui/display/toif.rs [hunk 3]
```diff
@@ -65,32 +50,45 @@ pub fn icon(icon: &Icon, center: Point, fg_color: Color, bg_color: Color) {
...
 #[derive(PartialEq, Eq, Clone, Copy)]
 pub struct Toif {
     data: &'static [u8],
 }
 
 impl Toif {
-    pub fn new(data: &'static [u8]) -> Self {
-        Self { data }
+    pub const fn new(data: &'static [u8]) -> Option<Self> {
+        if data.len() < TOIF_HEADER_LENGTH || data[0] != b'T' || data[1] != b'O' || data[2] != b'I'
```
Comment (matejcik) [inline core/embed/rust/src/ui/display/toif.rs:61]: `&data[0..3] != b"TOI"` ?

Comment (mmilata) [reply core/embed/rust/src/ui/display/toif.rs:61]: is not const

Comment (matejcik) [reply core/embed/rust/src/ui/display/toif.rs:61]: oh, makes sense.

**Review by matejcik — APPROVED** [review]

**Pull request merged**